src/coreclr/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs

```diff
@@ -826,5 +826,106 @@ public static IEnumerable<MethodDesc> EnumAllVirtualSlots(MetadataType type)
+                            continue;
+                        }
+
+                        if (!runtimeInterfaceType.HasSameTypeDefinition(interfaceType))
+                        {
+                            // Variance matches require a typedef match
+                            // Equivalence isn't sufficient, and is uninteresting as equivalent interfaces cannot have static virtuals.
+                            continue;
+                        }
+
+                        bool equivalentOrVariantCompatible;
+
+                        if (allowVariantMatches)
+                        {
+                            equivalentOrVariantCompatible = runtimeInterfaceType.CanCastTo(interfaceType);
+                        }
+                        else
+                        {
+                            // When performing override checking to ensure that a concrete type is valid, require the implementation 
+                            // actually implement the exact or equivalent interface.
+                            equivalentOrVariantCompatible = false; // TODO: runtimeInterfaceType.IsEquivalentTo(interfaceType);
+                        }
+
+                        if (equivalentOrVariantCompatible)
+                        {
+                            // Variant or equivalent matching interface found
+                            // Attempt to resolve on variance matched interface
+                            resolvedMethodOnType = TryResolveVirtualStaticMethodOnThisType(typeToCheck, runtimeInterfaceType, interfaceMethod);
+                            if (resolvedMethodOnType != null)
+                            {
+                                return resolvedMethodOnType;
+                            }
+                        }
+                    }
+                }
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Try to resolve a given virtual static interface method on a given constrained type and return the resolved method or null when not found.
+        /// </summary>
+        /// <param name="constrainedType">Type to attempt method resolution on</param>
+        /// <param name="interfaceType">Interface declaring the method</param>
+        /// <param name="interfaceMethod">Method to resolve</param>
+        /// <returns>MethodDesc of the resolved method or null when not found (runtime lookup must be used)</returns>
+        private static MethodDesc TryResolveVirtualStaticMethodOnThisType(TypeDesc constrainedType, TypeDesc interfaceType, MethodDesc interfaceMethod)
+        {
+            if (constrainedType is MetadataType mdType)
+            {
+                foreach (MethodImplRecord methodImpl in mdType.FindMethodsImplWithMatchingDeclName(interfaceMethod.Name) ?? Array.Empty<MethodImplRecord>())
+                {
+                    if (methodImpl.Decl.OwningType == interfaceType &&
+                        methodImpl.Decl == interfaceMethod)
```

Comment:
Do we actually need to pass the `interfaceType` around? This condition can be simplified to `if (methodImpl.Decl == interfaceMethod)` because the other part is redundant.

The CoreCLR type system is weird and needs to pass around both the method and the owning type separately, but we abstract that weirdness away in CorInfoImpl and don't import it into the general type system.

Reply:
Well, while I believe you're right that the outer method doesn't need the parameter, according to my understanding for <code>TryResolveVirtualStaticMethodOnThisType</code> we may need to match the method on a variance matched interface - perhaps that rather means that the method reference comparison is incorrect and we need to perform a structural comparison involving its name and signature?